### PR TITLE
implement squads for the backend

### PR DESCRIPTION
**Pull request by seanheinrichs**
### Core
- Create Squad Model
- Create Squad Route
- Create Squad Controller

### Secondary
- Slightly Changed All Sequelize Models, possibly the article you were using was a little old but all the methods you had declared are all built in (also had to change some of the methods).
- Moved sequelize object into its own database util instead of in the models file.
- Renamed all instances of `var` to `const` for imports. We should only be using const and let, var is dead from what I've read / been told.

## What's Missing?
- Tests (to come in the next PR)

### CODE REVIEW

File: backend/src/controllers/squad.js [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+  const userId = req.user.id;
+
+  try {
+    const currentUser = await req.context.models.User.findByPk(userId);
+    
+    // Create Squad
+    const squad = await currentUser.createSquad({
+      name: name,
+      memberCount: 1
+    });
```
Comment (mdesson) [inline backend/src/controllers/squad.js:54]: What's the idea behind creating the squad on the user? Membership isn't implemented, and sequelize isn't generating any new tables.

File: backend/src/controllers/squad.js [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+  const updatedName = req.body.name;
+
+  try {
+    // Fetch squad from database
+    const squad = await req.context.models.Squad.findByPk(req.params.squadId);
+
+    // Update squad if it exists
+    if (squad) {
+      squad.name = updatedName;
+      squad.save();
```
Comment (mdesson) [inline backend/src/controllers/squad.js:76]: Database operations are async, you should await in case there's an error. This way it doesn't return until it's successfully completed.

```suggestion
      await squad.save();
```

File: backend/src/controllers/squad.js [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+  }
+}
+
+exports.deleteSquad = async (req, res, next) => {
+  try {
+    // Fetch squad from database
+    const squad = await req.context.models.Squad.findByPk(req.params.squadId);
+  
+    // Delete squad from database
+    squad.destroy();
```
Comment (mdesson) [inline backend/src/controllers/squad.js:93]: Database operations are async, you should await in case there's an error. This way it doesn't return until it's successfully completed.

```suggestion
    await squad.destroy();
```

File: backend/src/controllers/squad.js [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+exports.getSquads = async (req, res, next) => {
+  try {
+    // Fetch squads from database
+    const squads = await req.context.models.Squad.findAll({ where: { userId: req.user.id }});
+    
+    // Return No Content if the user has no squads
+    if (squads.length < 1) {
+      res.status(204);
+      return res.send({ message: "This user has no squads."})
+    }
```
Comment (mdesson) [inline backend/src/controllers/squad.js:26]: The database query and the error message are two different things here:

What you're checking is if the current user _created_ the squad, since that's all you're recording.

The error message makes it sounds like the user is not a _member_ of any squads, which isn't implemented yet.

File: backend/src/controllers/squad.js [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+    const squad = await req.context.models.Squad.findByPk(req.params.squadId);
+    const squadData = squad.dataValues;
+  
+    // Return squad
+    res.status(201);
+    res.send(squadData);
+  } catch (err) {
+    console.log(err)
+    res.status(404);
+    res.send({ error: "Could not find squad." });
```
Comment (mdesson) [inline backend/src/controllers/squad.js:13]: I wasn't doing this either, but to make it a little cleaner we could fit these onto one line ([source](https://stackoverflow.com/questions/26066785/proper-way-to-set-response-status-and-json-content-in-a-rest-api-made-with-nodej))

```suggestion
    res.status(404).send({ error: "Could not find squad." });
```

File: backend/src/middleware/appContext.js [hunk 6]
```diff
@@ -1,6 +1,8 @@
-var { models } = require("../models");
+const models = require("../models");
 
 module.exports = async (req, res, next) => {
-  req.context = { models };
+  req.context = {
+    models: models
+  };
```
Comment (mdesson) [inline backend/src/middleware/appContext.js:6]: What's the purpose of this change? This works just as well:

```suggestion
  req.context = { models };
```
This way you can require as needed, and toss it into the context directly

File: backend/src/models/index.js [hunk 7]
```diff
@@ -1,22 +1,16 @@
...
+const Auth = require('./auth');
+const Squad = require('./squad');
 
 const models = {
-  User: sequelize.import("./user"),
-  Auth: sequelize.import("./auth"),
+  User: User,
+  Auth: Auth,
+  Squad: Squad,
 };
```
Comment (mdesson) [inline backend/src/models/index.js:9]: It would be a little cleaner as:

```js
const models = { User, Auth, Squad };
```

File: backend/src/routes/index.js [hunk 8]
```diff
@@ -1,5 +1,7 @@
-var user = require("./user");
+const user = require("./user");
+const squad = require("./squad");
 
 module.exports = {
   user,
+  squad,
 };
```
Comment (mdesson) [inline backend/src/routes/index.js:7]: Slightly cleaner:

```suggestion
module.exports = { user, squad, };
```

File: backend/src/controllers/squad.js (hunk 1 above)
Comment (seanheinrichs) [reply backend/src/controllers/squad.js:54]: This isn't using sequelize to it's full potential though... https://sequelize.readthedocs.io/en/v3/api/associations/has-many/

The whole point of these association methods is that it auto builds them for you. In our example, a user HAS MANY squads. So a user can createSquad() and sequelize automatically connects the table for you.

When we start working with many : many relationships I would much rather have sequelize handle all the relation mapping then trying to hardcode it ourselves. 

Can you read the documentation and let me know if you're still not sold? Maybe try it with user and auth: https://sequelize.readthedocs.io/en/v3/api/associations/has-one/

Comment (mdesson) [reply backend/src/controllers/squad.js:54]: All good points! This works for me! I'll create a ticket to match auth/user to how you have it, for the 1:1. 

> Users and Squads are 1 : m not 1:1 unless I'm misunderstanding.

To clarify, I was referring to what's currently implemented in this pr. When you look in the database, all it's adding `userId` to the squads table. Which would represent a 1:1 relation (representing the creator of the squad). The way it's implemented currently there's no 1:m relation as far as I can see and test. 

As a sidenote, squads/users is m:m. One user has many squads. One squad has many users. No use or squad is uniquely tied to the other.

Comment (seanheinrichs) [reply backend/src/controllers/squad.js:54]: hmmmmm, maybe I'm not understanding but what I was reading says otherwise: https://code.tutsplus.com/articles/sql-for-beginners-part-3-database-relationships--net-8561

There's an example of 1 to many where 1 customer has many orders. This is like one user being able to create many squads. no?

Comment (mdesson) [reply backend/src/controllers/squad.js:54]: Correct, my bad!

File: backend/src/middleware/appContext.js (hunk 6 above)
Comment (mdesson) [reply backend/src/middleware/appContext.js:6]: As a sidenote, I'm going to make a ticket for a small task to move `req.user` to `req.context.user`. The purpose of context is to provide app context for the request. `req.user` makes it sound like it's part of the request, not the context surrounding it.

That or we can drop context, since it's pretty verbose for not much. We can discuss it.
